### PR TITLE
use k8s resource status to decide resource health

### PR DESCRIPTION
**Pull request by nimakaviani**
- check status from cluster resources and if they report
Ready or Available, take the value as an indication of resource health
- refactor resolvers so the evaluation and event emission is only done once

### CODE REVIEW

**Comment by nimakaviani** [comment]
It has a dependency on `refactor-k8s-manifest` so needs to wait for the other one to get merged first.

**Comment by nimakaviani** [comment]
This PR is solving for issue: https://github.com/nimakaviani/managed-delivery-k8s-plugin/issues/37

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt [hunk 1]
```diff
@@ -13,13 +13,34 @@ data class K8sData(
...
+        if (it.type == "Ready" || it.type == "Available")
+            return it.status == "True"
+    }
+    return true
+}
+
 open class K8sManifest(
-    open val apiVersion: String?,
-    open val kind: String?,
+    open var apiVersion: String?,
```
Comment (nabuskey) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:37]: Why do these need to be `var`?  Because of `jacksonObjectMapper` ?

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt [hunk 2]
```diff
@@ -65,6 +67,22 @@ abstract class GenericK8sResourceHandler <S: GenericK8sLocatable, R: K8sManifest
...
+                    eventPublisher.publishEvent(ResourceHealthEvent(r, true))
+                } else {
+                    log.info("${r.spec.displayName} is not healthy")
+                    eventPublisher.publishEvent(ResourceHealthEvent(r, false))
+                    throw ResourceNotReady(r)
+                }
+            }
+            it
+        }
+
```
Comment (nabuskey) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt:85]: We should ensure we return `ResourceNotReady` or null when clouddriver returns 404. Otherwise we are going to be filling keel log with 404 exceptions due to user error. Also I think it is possible for clouddriver to return 404 especially when it's still being provisioned.

**Review by nabuskey — APPROVED** [review]

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt (hunk 1 above)
Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:37]: we rewrite values for `metadata`, `spec`, etc in the tests to check corner cases. so all attribtues except for `apiVersion` and `kind` were already variable attributes. I just changed these for consistency.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt (hunk 2 above)
Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt:85]: isnt that whats happening already?

Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt:85]: oh, did you mean either or by this `ensure we return ResourceNotReady or null when clouddriver returns 404`?

we are returning null, which cascades to `null` in the subclasses as well.

Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt:85]: So if there are any errors during `getK8sResource(r)` , it's going to throw an exception. It's not an issue if implementation of  `getK8sResource(r)` caught specific errors and returned null. I am not seeing that in any of implementations.

Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt:85]: yeah so 404 return null, anything else throws an exception which bubbles up and sets the status of the resource to `ERROR`. why do we want to catch the exception? The exception will go all the way to Keel for it to be caught and reported.
